clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -548,6 +550,13 @@ public void handle(JoinGroupResponse joinResponse, RequestFuture<ByteBuffer> fut
                 future.raise(error);
             } else if (error == Errors.GROUP_AUTHORIZATION_FAILED) {
                 future.raise(new GroupAuthorizationException(groupId));
+            } else if (error == Errors.MEMBER_ID_REQUIRED) {
+                // Broker requires a concrete member id to be allowed to join the group. Update generation and member id
+                // and send another join group request in next cycle.
+                AbstractCoordinator.this.generation = new Generation(joinResponse.generationId(),
```

Comment:
I'm slightly uncomfortable with the additional complexity that comes from having an incomplete `Generation` object. At a minimum, we probably need to change the checks which rely on `Generation.NO_GENERATION`. For example, see `maybeLeaveGroup`. Perhaps we can add an `isValid` method to `Generation` which only returns true if the generationId is valid?